### DISCUSSION TITLE
Get Computer Host IP address from Micropython Board

### DISCUSSION THREAD

**Discussion by Danilux** [micropython/micropython#15679] [General]
Is there a way to get the computer host ip address from the micropython board (ESP8266), the board is plugged in to the computer via USB, they are both connected to the same wifi network, ideally the solution would be from the board side only, not having to prepare a script to send the ip from the computer to the board.

**Comment by glenn20** [comment]
I don't believe that is possible. You don't want the board having the ability to execute arbitrary code on your host computer.

You would need to run a script on your host computer (whatever OS you are running) to get the correct IP address and copy it onto the board (eg. with mpremote).

**Reply by Josverl** [reply]
Fully agree, such a possibility would constitute a significant security vulnerability

**Reply by Danilux** [reply]
Could I send the ip address from the host pc using python through usb, serial communication or wifi? are there other ways on micropython

**Reply by andypiper** [reply]
yes, you could do that if you wanted, you would need to run it directly on the host system as mentioned  in the original response.

**Reply by bixb922** [reply]
> Could I send the ip address from the host pc using python through usb, serial communication or wifi? are there other ways on micropython

One simple way to do that is using ```mpremote exec``` on the PC at command level or in a shell script

**Reply by Danilux** [reply]
I'm going to test that as proof of concept, but mi final solution will need to be more complex for ease of use for users, I'm thinking of creating a website with form controls selections that creates a python config file and using something like websocket send this file and upload it to the esp8266, I have no docs where to start or similar projects you can point is appreciated

**Reply by Josverl** [reply]
Can you describe what your intended solution would do, and why it needs one of the host's IP addresses, and how it would use that?

**Reply by peterhinch** [reply]
The phrase "host's IP" - which IP are we talking about? The IP of the PC on the LAN, or the IP assigned by the ISP to your internet connection?

**Reply by Danilux** [reply]
> Can you describe what your intended solution would do, and why it needs one of the host's IP addresses, and how it would use that?

it's a simple esp2688 keypad, used micropython to send network commands to the PC the board is plugged in on key press, all that is working great already, the part that I need to build now is a webapp (website?) that would let a user configure (change) each key command from a web form, and send those changes to the esp2688 to update the variables on the board, I want this to be done via the same usb wire through serial (?), not wifi, I'm a web programmer so the website part is easy, the sending data, updating variables, and saving them on the board have no clue yet how to do it.

**Reply by Danilux** [reply]
> n the LAN, or the IP assigned

Sorry, the ip of the computer the board is plugged in, the one on the LAN.
Not the one assigned to the router by the ISP.

**Reply by Josverl** [reply]
Then I'd suggest something like
- configure the esp8266 with a name such as `keypad`
- running the website on the esp using microdot or similar
  - the static content can actually be stored elsewhere (GitHub pages, Azure, ec3 ....) to lighten the load
   - configuration will be POST-ed back to the MCU, which can save it
- that allows the host to browse to `http://keypad.local` resolving using mdns

If needed you can check the source IP of the http GET/POST messages.
That may or may not be the host's local IP (IPv4/V6, proxy servers, vpns, virtualization, docker, .....) , but I don't think you'll need it anyway 

HtH

**Comment by bulletmark** [comment]
I'd run the webapp on your host computer (where it is much easier to develop and you have much better packages, tools, etc) and store all configuration there also to keep the MP device end as simple as possible. Don't need anything on the device other than stock MP. E.g. your host software can watch the serial/usb ports and see when a MP device is connected then `mpremote run` your code (i.e. run on the device but loaded directly from your host). The first thing that code would do is check the device MAC address and then choose the appropriate configuration for that node, assuming you will have potentially multiple of these devices connecting to your host on different USB ports. The web app can communicate config changes to the app or possibly just restart the `mpremote run` on a config change? Development cycle would be extremely fast with this approach.
